### PR TITLE
Button handling

### PR DESCRIPTION
**Pull request by xasz**
Added Methods for Buttons like Game Engine do:

` // Buttons (Engine-Style)
  bool btnHasGoneDown(uint8_t btn);
  bool btnHasGoneUp(uint8_t btn);
  bool btnIsPhysicalDown(uint8_t btn);
  bool btnIsLongPress(uint8_t btn);`

U can now easy acess events in your app-loop like:

`if(hal->btnHasGoneDown(2)){
  startMyAction();
}`

Added a small app to test the events.

Issue: #67

### CODE REVIEW

File: include/osw_hal.h [hunk 1]
```diff
@@ -142,6 +149,13 @@ class OswHal {
   long _btn1UpSince = 0;
   long _btn2UpSince = 0;
   long _btn3UpSince = 0;
+  bool btnLastState[3];
+  bool btnIsDown[3];
+  bool btnGoneUp[3];  
+  bool btnSuppressUntilUpAgain[3];  
+  bool btnGoneDown[3];   
+  long btnIsDownSince[3];
```
Comment (SeanReg) [inline include/osw_hal.h:157]: Should be `unsigned long`

File: src/hal/buttons.cpp [hunk 2]
```diff
@@ -46,12 +62,53 @@ void OswHal::checkButtons(void) {
...
+  btnIsDown[2] = !digitalRead(BTN_3) == LOW;
+
+  for(uint8_t i = 0  ; i < 3 ; i++){
+    btnGoneUp[i] = btnLastState[i] == true && btnIsDown[i] == false;
+    btnGoneDown[i] = btnLastState[i] == false && btnIsDown[i] == true;
+    
+    if(btnGoneDown[i]){
+      btnIsDownSince[i] = millis();  
+    }
+    btnLongPress[i] = btnIsDownSince[i] < millis() - DEFAULTLAUNCHER_LONG_PRESS && btnIsDown[i] == true;
```
Comment (SeanReg) [inline src/hal/buttons.cpp:78]: `millis() > btnIsDownSince[i] + DEFAULTLAUNCHER_LONG_PRESS` would protect against overflow issues

**Comment by SeanReg** [comment]
Some sort of input map would help here. Such as:

`
enum Input {
   INPUT_MIN

    BUTTON_1,
    BUTTON_2,
    BUTTON_3,

   INPUT_MAX
}

//Map inputs to pins
InputMap = {
   BTN_1,
   BTN_2,
   BTN_3
}
`

You could then setup the pin modes and iterate over the buttons using these. Something like:

`
for (unsigned int in = INPUT_MIN + 1; in != INPUT_MAX; ++in) {
  unsigned int idx= in - 1;
  Input inputId = static_cast<Input>(in);

  digitalRead(InputMap[idx])
}
`

This would remove a lot of the harcoded numbers and allow future buttons or inputs to more easily be added in the future if ever needed.

File: src/hal/buttons.cpp (hunk 2 above)
Comment (xasz) [reply src/hal/buttons.cpp:78]: Yeah we can chance that, when its switched to unisgned :) Will chance that.

**Comment by xasz** [comment]
@SeanReg Thanks for the review and the idea is great.
I thought about some enum aswell but i was not sure. I had a talk with Paul yesterday about #55 and  he wants to keep stuff as simple as possible and i am fine with that. Thats why i did not use any castings and make it simple to understand.

What could be a solution/compromiss would be some simple defines for the buttons which then can be used in the code like:

`
  #define BUTTON_1 = 0
  #define BUTTON_2 = 1
  #define BUTTON_3 = 2
`

@uvwxy What would u like to have implemented ?

**Comment by uvwxy** [comment]
The enum is fine 👍

**Comment by xasz** [comment]
Perfect.

**Comment by xasz** [comment]
i added a new commit which hope does make happy everyone for now :)

**Pull request merged**